### PR TITLE
Use `ignore_conflicts` when inserting references

### PR DESCRIPTION
**Pull request by chris48s**
This PR fixes a race condition which can occur when inserting into the `ReferenceIndex`.

`create_or_update_for_object()` attempts to do this safely and the docs say

> This method must be called within a `django.db.transaction.atomic()` block.

but fundamentally, it is still possible to race it.

We run a select here
https://github.com/wagtail/wagtail/blob/953c980976231d6d2e1d26a58595710dcf10cc71/wagtail/models/reference_index.py#L439-L454

The we use that to determine what we're going to insert and do a `bulk_create` here:

https://github.com/wagtail/wagtail/blob/953c980976231d6d2e1d26a58595710dcf10cc71/wagtail/models/reference_index.py#L460-L474

If some other process inserts a record with the same `(base_content_type_id, object_id, to_content_type_id, to_object_id, content_path_hash)` as one of the records we plan to insert inbetween us doing that select and attempting the insert, we will throw an error like

```
IntegrityError

duplicate key value violates unique constraint "wagtailcore_referenceind_base_content_type_id_obj_9e6ccd6a_uniq"
DETAIL:  Key (base_content_type_id, object_id, to_content_type_id, to_object_id, content_path_hash)=(125, 850, 133, 4, 8d8f00fb-91d7-5c14-a491-1624ddbe1421) already exists.
```

`transaction.atomic()` won't save us from this.

This is quite hard to write a test case for as it is a concurrency problem.

My proposed solution is to use `ignore_conflicts=True`. I think based on my (rapidly acquired) understanding of the `ReferenceIndex` this should not actually cause data loss as two `ReferenceIndex` objects with the same `(base_content_type_id, object_id, to_content_type_id, to_object_id, content_path_hash)` would implicitly have the same other fields anyway?

The other options I considered were:

- `update_conflicts`: I discounted this because it is available in Django 4.1+ and django down to 3.2 will still be supported in Wagtail 5.2. `ignore_conflicts` is available in django 3.2
- Completely locking the `ReferenceIndex` table and disallowing dirty reads while `create_or_update_for_object()` is running. This would introduce an unacceptable performance bottleneck IMO. A fairly large subset of write operations are going to call this function.

Also, as I say, I _think_ `ignore_conflicts` will serve our needs in this case anyway.

It is worth noting that `ignore_conflicts` has some MySQL/MariaDB-specific notes in https://docs.djangoproject.com/en/3.2/ref/models/querysets/#bulk-create

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

Basically you need to kick off a bunch of requests that all want to insert the same object into the reference index at the same time. In our application, this manifests when the frontend launches multiple POST requests to different API endpoints in a `Promise.all()`. That would probably be the easiest way to construct a toy repro.

### CODE REVIEW

**Comment by laymonage** [comment]
Worth noting that `ignore_conflicts` is not supported on MSSQL. We don't officially support MSSQL, but there have been cases where we decided to avoid using such features e.g. #10836. 

That said, in there specifically, the code is run every time `migrate` is run, rendering the project unusable on MSSQL. The one in this PR is only run when a reference is recorded, so it's less severe but it means the reference index won't be usable on MSSQL.

**Comment by chris48s** [comment]
Another option would be to do something like

```py
from django.db import connection

if connection.vendor == 'microsoft':
    bulk_create_kwargs = {}
else:
    bulk_create_kwargs = {'ignore_conflicts': True}

cls.objects.bulk_create(..., **bulk_create_kwargs)
```

so the code is executable if you use SQL server (but can still be raced) and fix it for the supported engines?

**Comment by jacobtoppm** [comment]
I think that would be a neat solution. It would be nice not to break MSSQL, but given it's not a supported backend, and the race condition is causing issues on supported backends, I think not breaking MSSQL, but actually fixing the issue only for the supported backends, feels like enough for now.

File: wagtail/models/reference_index.py [hunk 1]
```diff
@@ -456,6 +456,14 @@ def create_or_update_for_object(cls, object):
         # Construct the set of reference records that have been found on the object but are not
         # already present in the database
         new_references = references - set(existing_references.keys())
+
+        # ignore_conflicts is not available on MSSQL server
+        # refs https://github.com/wagtail/wagtail/pull/11025#issuecomment-1755253875
+        if connection.vendor == "microsoft":
```
Comment (laymonage) [inline wagtail/models/reference_index.py:462]: Let's use Django's `DatabaseFeatures`, this is how they and third party packages tend to check for feature support:

https://github.com/django/django/blob/e083f3082c71853a01bf149bda7fdbaf58d25f4d/django/db/backends/base/features.py#L293-L295

```suggestion
        if connection.features.supports_ignore_conflicts:
```

Comment (chris48s) [reply wagtail/models/reference_index.py]: ooh nice. TIL :mortar_board:

**Review by jacobtoppm — APPROVED** [review]

**Comment by jacobtoppm** [comment]
Merged in https://github.com/wagtail/wagtail/commit/4219488a3ce93642781e5be2eda938718380d40d
